lychee-lib/src/extract/markdown.rs

```diff
@@ -77,6 +81,49 @@ pub(crate) fn extract_markdown(input: &str, include_verbatim: bool) -> Vec<RawUr
         .collect()
 }
 
+/// Extract unparsed URL strings from a Markdown string.
+pub(crate) fn extract_markdown_fragments(input: &str) -> HashSet<String> {
+    let mut in_heading = false;
+    let mut heading = String::new();
+    let mut kebab_id_counter = HashMap::<String, usize>::new();
+
+    let mut out = HashSet::new();
+
+    for event in Parser::new(input) {
+        match event {
+            Event::Start(Tag::Heading(..)) => {
+                in_heading = true;
+            }
+            Event::End(Tag::Heading(_d, id, ..)) => {
+                if let Some(frag) = id {
+                    out.insert(frag.to_string());
+                }
+
+                let mut id = heading.to_case(Case::Kebab);
+
+                let k_id_count = kebab_id_counter.entry(id.clone()).or_insert(0);
+                if *k_id_count != 0 {
+                    id = format!("{id}-{k_id_count}");
+                }
+                *k_id_count += 1;
```

Comment:
Maybe we can extract that id-generation logic into a separate struct:

```rust
struct HeadingIdGenerator {
    counter: HashMap<String, usize>,
}

impl HeadingIdGenerator {
    fn new() -> Self {
        Self {
            counter: HashMap::new(),
        }
    }

    fn generate(&mut self, heading: &mut String) -> String {
        let mut id = heading.to_case(Case::Kebab);
        let count = self.counter.entry(id.clone()).or_insert(0);
        if *count != 0 {
            id = format!("{}-{}", id, *count);
        }
        *count += 1;

        id
    }
}
```

Might look a bit like overkill, but it's easier to test, separates the logic, and keeps track of state.

You can then initialize it at the beginning of this function like so:

```rust
let mut id_generator = HeadingIdGenerator::new();
```

and use it like

```rust
let id = id_generator.generate(&mut heading);
out.insert(id);
```

This gives us full control over how IDs get generated. What do you think?

Reply:
I have no strong opinion either way, so it now exists.